### DISCUSSION TITLE
Is there any inheritance or metaclass magic for less-repetitive Admin class definitions?

### DISCUSSION THREAD

**Discussion by afparsons** [aminalaee/sqladmin#223] [General]
When defining various Admin classes, my code is quite [WET](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself):

```python
class FooAdmin(ModelAdmin, model=Foo):
    column_list = Foo.__table__.columns.keys()
    column_details_list = Foo.__table__.columns.keys()
    form_columns = Foo.__table__.columns.keys()

class BarAdmin(ModelAdmin, model=Bar):
    column_list = Bar.__table__.columns.keys()
    column_details_list = Bar.__table__.columns.keys()
    form_columns = Bar.__table__.columns.keys()
```

I was hoping to make it [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) through some kind of inheritance:

```python
class MixinAllColumns:
    """
    Obviously, this won't work, but just consider the idea
    """
    column_list = model.__table__.columns.keys()
    column_details_list = model.__table__.columns.keys()
    form_columns = model.__table__.columns.keys()

class FooAdmin(ModelAdmin, MixinAllColumns, model=Foo):
    """
    """

class BarAdmin(ModelAdmin, MixinAllColumns, model=Bar):
    """
    """
```
It is not obvious to me how (or even _if_) `MixinAllColumns` could get access to `model`. 

Does anyone have any ideas?

_Nota bene:_ omitting `column_details_list` and `form_columns` (I believe they default to all the columns anyway) results in an error:

```
sqlalchemy.orm.exc.DetachedInstanceError: Parent instance <RulePattern at 0x7f6a002d9100> is not bound to a Session; lazy load operation of attribute 'rules' cannot proceed (Background on this error at: https://sqlalche.me/e/14/bhk3)
```

...hence why I am manually defining them.

This is likely a general Python (metaclass) question and less related to sqladmin itself. Otherwise, great work on sqladmin!

**Comment by aminalaee** [comment]
Hey, I think this might be very common and it should be possible, it's how the sqladmin is working. But I'll take a look later to see how this can be done and maybe add it to the docs as it can be very common.

**Comment by aminalaee** [comment]
BTW I think your use-case is that you want to always show all columns in list page, I think that's what flask-admin does and I didn't.

Maybe it's better to stick to what they have done.

**Reply by afparsons** [reply]
Thanks for fielding my question so quickly!

> BTW I think your use-case is that you want to always show all columns in list page

Yes, exactly. I *could* manually define all the columns with string names, but I was hoping to keep the column list dynamic (one less thing for me to update if I change the table schema).

And again, I get a `sqlalchemy.orm.exc.DetachedInstanceError` if I define `column_list` but omit `column_details_list` and `form_columns`.

**Reply by aminalaee** [reply]
@afparsons I haven't had a chance to take a look into it.
If you're interested I think we can change this behaviour inside SQLAdmin so that all columns are shown by default, as I mentioned this is also the default behaviour in Flask-Admin
